src/replication.c

```diff
@@ -2071,14 +2368,164 @@ handle_update_spec_stats(spec_t *spec, zvol_io_hdr_t *hdr, void *resp)
+ * This operation must be performed by taking spec lock and pause IOs
+ * until timeout or operation completes.
+ */
+static int
+update_replication_factor(spec_t *spec, replica_t *replica) {
+	int known_replica_count, data_len;
+	char *data;
+	const char *json_string;
+	struct json_object *jobj;
+	int rf, cf, rc = -1;
+	bool old_replica;
+
+	ASSERT(MTX_LOCKED(&spec->rq_mtx));
+
+	known_replica_count = get_known_replica_count(spec);
+	old_replica = is_replica_exist(spec, replica->replica_id);
+
+	/* If is_replica_exist return true then it is a kind of replica replace
+	 * or replica movement scenarios. If it return false then it is a replica
+	 * scale up case then to add replica into known replica list current known 
+	 * replica count should be less than desired replication factor.
+	 */
+	if (!old_replica && known_replica_count >= spec->desired_replication_factor) {
+		ISTGT_ERRLOG("failed to update replication/consistency factor "
+		    "known replica count %d is greater than or equal to "
+		    "desired replciation factor %d\n",
+		    known_replica_count, spec->desired_replication_factor);
+		return rc;
+	}
+
+	ASSERT(spec->quiesce == 1);
+	MTX_LOCK(&replica->r_mtx);
+
+	/* Update RF, CF and known replica list*/
+	rf = spec->replication_factor + 1;
+	cf = (rf/2) + 1;
+	jobj = json_object_new_object();
+	json_object_object_add(jobj, "replicationFactor",
+	    json_object_new_int(rf));
+	json_object_object_add(jobj, "consistencyFactor",
+	    json_object_new_int(cf));
+	json_object_object_add(jobj, "volumeName",
+	    json_object_new_string(spec->volname));
+	json_object_object_add(jobj, "replicaId",
+	    json_object_new_string(replica->replica_id));
+	json_object_object_add(jobj, "replicaZvolGuid",
+	    json_object_new_uint64(replica->zvol_guid));
+	json_string = json_object_to_json_string_ext(jobj, JSON_C_TO_STRING_PLAIN);
+
+	data_len = strlen(json_string) + 1;
+	data = (char *) malloc(sizeof(char *)*data_len);
+	memset(data, 0, data_len);
+	strncpy(data, json_string, data_len);
+
+	ISTGT_LOG("updating non-quorum replica(%s:%lu) to known replica list\n",
+	    replica->replica_id, replica->zvol_guid);
+	rc = update_volume_config(data);
+	if (rc < 0) {
+		MTX_UNLOCK(&replica->r_mtx);
+		ISTGT_ERRLOG("failed to update replica(%s:%lu) to known "
+		    "replica list\n", replica->replica_id, replica->zvol_guid);
+		free(data);
+		json_object_put(jobj);
+		return rc;
+	}
+
+	/* Update inmemory RF and CF */
+	if (!old_replica) {
+		ISTGT_LOG("Updated the replication_factor: from %d to %d "
+		    ",consistency_factor: from %d to %d and replica known list {%lu}\n",
+		    spec->replication_factor, rf, spec->consistency_factor,
+		    cf, replica->zvol_guid);
+		spec->replication_factor = rf;
+		spec->consistency_factor = cf;
+		spec->lu->replication_factor = rf;
+		spec->lu->consistency_factor = cf;
+	}
+	(void) update_in_memory_known_replica_list(spec, replica);
+	rc = 1;
+	MTX_UNLOCK(&replica->r_mtx);
```

Comment:
why we need this lock?

Reply:
we are accessing fields in replica so took the lock